### ISSUE TITLE
Create new contacts-BE

### ISSUE THREAD

**Issue by epintozzi** [turingschool/tracker-crm#1]
12-9-24 All action items are complete for the BE. Will reference this ticket in PR--which will be done by Renee for her ticket as we were on the same branch feat/contacts-controller-rm-cc.

Logged in users should be able to add new contacts which are only visible to the logged in user (we don't want to expose a person's contacts to other people)

Fields:

- *First Name
- *Last Name
- Company (dropdown that links to existing company in database)
- Email
- Phone Number
- Notes

`*` Indicates required

![Image](https://github.com/user-attachments/assets/4be1c9da-e162-4636-b785-0da2c0d54065)

**Comment by CCirbo** [comment]
- [ ] Create a Contacts table with these items
 Fields:
        - *First Name
        - *Last Name
        - Company (dropdown that links to existing company in database)
        - Email
        - Phone Number
        - Notes

**Comment by CCirbo** [comment]
- [ ] Route for create action
- [ ] Controller for create action
- [ ] Rspec tests for controller create action
- [ ] Rspec tests for model validations
- [ ] Additional migrations for Contacts table--fix company from string placeholder to foreign key/company id
- [ ] Add regex code for phone number
- [ ] Add regex code for email
- [ ] Rspec tests for model methods/code for phone number and email
- [ ] Update associations for company id, phone number and email -- optional in model

**Comment by epintozzi** [comment]
Candice, can you also put the JSON contract (expected request/response) here in the issue? It looks like the route and controller are already done so hopefully that's easy to whip up and document here. Ideally, that would be documented before the work in case anyone needs to pick it up from you unexpectedly.

**Comment by CCirbo** [comment]
Yes, I was actually doing that in postman now, I got a little enthusiastic about my check marking off boxes!

**Comment by CCirbo** [comment]
In Postman to get to create contact you have to get your sessions token first.
Go to
POST http://localhost:3001/api/v1/sessions
  In body, put in this or any email and password of user
  {
    "email": "123@email.com",
    "password": "Abc123"
}

This gets you a token: dkjkdjlsdjf.........

Then do a new request:
POST http://localhost:3001/api/v1/contacts

Go to Authorization tab, Auth Type and scroll down to Bearer Token
add token to Token Box

Go to Body
add this or whatever you want for first and last name
  {
    "first_name": "Danny",
    "last_name": "Jersey"
}

JSON response to new contact creation. All attributes other than first and last name are optional.

{
    "data": {
        "id": "3",
        "type": "contacts",
        "attributes": {
            "first_name": "Danny",
            "last_name": "Jersey",
            "company": null,
            "email": null,
            "phone_number": null,
            "notes": null,
            "user_id": 4
        }
    }
}

**Comment by CCirbo** [comment]
While implementing the create action, I identified an issue where a user could create multiple contacts with the same first and last name. To address this, I added validation to the model and database to ensure that each user can only create one unique contact with the same first and last name. This validation is scoped to each individual user, allowing different users to have contacts with the same name. The limitation applies only to each user and does not affect all users globally.
I also tested this in the contacts model spec file.
- [ ] Added unique name validation to model
- [ ] Added unique name validation to contacts table via migration
- [ ] Rspec tests to contacts_spec file for additional validation
